openfe/protocols/restraint_utils/geometry/boresch.py

```diff
@@ -0,0 +1,870 @@
+
+    Where HX represents the X index of ``host_atoms`` and GX
+    the X index of ``guest_atoms``.
+    """
+    r_aA0: FloatQuantity['nanometer']
+    """
+    The equilibrium distance between H0 and G0.
+    """
+    theta_A0: FloatQuantity['radians']
+    """
+    The equilibrium angle value between H1, H0, and G0.
+    """
+    theta_B0: FloatQuantity['radians']
+    """
+    The equilibrium angle value between H0, G0, and G1.
+    """
+    phi_A0: FloatQuantity['radians']
+    """
+    The equilibrium dihedral value between H2, H1, H0, and G0.
+    """
+    phi_B0: FloatQuantity['radians']
+
+    """
+    The equilibrium dihedral value between H1, H0, G0, and G1.
+    """
+    phi_C0: FloatQuantity['radians']
+
+    """
+    The equilibrium dihedral value between H0, G0, G1, and G2.
+    """
+
+
+def _sort_by_distance_from_atom(
+    rdmol: Chem.Mol, target_idx: int, atom_idxs: Iterable[int]
+) -> list[int]:
+    """
+    Sort a list of RDMol atoms by their distance from a target atom.
+
+    Parameters
+    ----------
+    target_idx : int
+      The idx of the atom to measure from.
+    atom_idxs : list[int]
+      The idx values of the atoms to sort.
+    rdmol : Chem.Mol
+      RDKit Molecule the atoms belong to
+
+    Returns
+    -------
+    list[int]
+      The input atom idxs sorted by their distance from the target atom.
+    """
+    distances = []
+
+    conformer = rdmol.GetConformer()
+    # Get the target atom position
+    target_pos = conformer.GetAtomPosition(target_idx)
+
+    for idx in atom_idxs:
+        pos = conformer.GetAtomPosition(idx)
+        distances.append(((target_pos - pos).Length(), idx))
+
+    return [i[1] for i in sorted(distances)]
+
+
+def _bonded_angles_from_pool(
+    rdmol: Chem.Mol, atom_idx: int, atom_pool: list[int]
+) -> list[tuple[int, int, int]]:
+    """
+    Get all bonded angles starting from ``atom_idx`` from a pool of atoms.
+
+    Parameters
+    ----------
+    rdmol : Chem.Mol
+      The RDKit Molecule
+    atom_idx : int
+      The index of the atom to search angles from.
+    atom_pool : list[int]
+      The list of indices to pick possible angle partners from.
+
+    Returns
+    -------
+    list[tuple[int, int, int]]
+      A list of tuples containing all the angles.
+    """
+    angles = []
+
+    # Get the base atom and its neighbors
+    at1 = rdmol.GetAtomWithIdx(atom_idx)
+    at1_neighbors = [at.GetIdx() for at in at1.GetNeighbors()]
+
+    # We loop at2 and at3 through the sorted atom_pool in order to get
+    # a list of angles in the branch that are sorted by how close the atoms
+    # are from the central atom
+    for at2 in atom_pool:
+        if at2 in at1_neighbors:
+            at2_neighbors = [
+                at.GetIdx() for at in rdmol.GetAtomWithIdx(at2).GetNeighbors()
+            ]
+            for at3 in atom_pool:
+                if at3 != atom_idx and at3 in at2_neighbors:
+                    angles.append((atom_idx, at2, at3))
```

Comment:
From today's discussion: we need to avoid cases where we can cross rings with a non-aromatic bond. Most likely easiest way to do this is to have a special case here that can optionally filter by ring (i.e. only pick angles where you have atoms indices that exist solely within one ring).